File: lib/System/IO/Tun.hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE QuasiQuotes     #-}
module System.IO.Tun
    ( openTun
    ) where

import Foreign.C.String (CString, withCString)
import Language.C.Inline (CInt)
import System.Posix.Types (Fd (..))

import qualified Language.C.Inline as C

C.include "<stdio.h>"
C.include "<string.h>"
C.include "<unistd.h>"
C.include "<fcntl.h>"
C.include "<sys/ioctl.h>"
C.include "<net/if.h>"
C.include "<linux/if_tun.h>"

openTun :: String -> IO (Maybe Fd)
openTun path = do
    fd <- withCString path openTunC
    if fd < 0 then return Nothing
              else return $ Just (Fd fd)

openTunC :: CString -> IO CInt
openTunC path =
    [C.block| int {
        struct ifreq ifr;
        int fd = -1;

        printf("Got %s\n", $(char* path));
        return fd;
    }|]

File: lib/System/IO/Tun.c

#include <stdio.h>

#include <string.h>

#include <unistd.h>

#include <fcntl.h>

#include <sys/ioctl.h>

#include <net/if.h>

#include <linux/if_tun.h>

int inline_c_System_IO_Tun_0_bd2a7fb9efe496bd2ce0895e72f4562981ffe9d7(char * path_inline_c_0) {

        struct ifreq ifr;
        int fd = -1;

        printf("Got %s\n", path_inline_c_0);
        return fd;
    
}

